src/main/java/org/jboss/modules/ModuleLoggerFinder.java

```diff
@@ -34,64 +43,175 @@
+     * from the new finder.
      *
      * @param cl the class loader to use
      */
-    static void setClassLoader(final ClassLoader cl) {
-        classLoader = cl;
-    }
-
-    private LoggerFinder resolveFinder() {
-        LoggerFinder result = finder;
-        if (result == null) {
-            synchronized (this) {
-                result = finder;
-                if (result == null) {
-                    try {
-                        final ClassLoader cl = classLoader;
-                        final ServiceLoader<LoggerFinder> loader = ServiceLoader.load(LoggerFinder.class, cl == null ? ClassLoader.getSystemClassLoader() : cl);
-                        for (LoggerFinder lf : loader) {
-                            if (!getClass().equals(lf.getClass())) {
-                                finder = result = lf;
-                                break;
-                            }
+    static void activate(final ClassLoader cl) {
+        // To keep order, we must observe the lock
+        lock.lock();
+        try {
+            if (!activated) {
+                try {
+                    final ServiceLoader<LoggerFinder> loader = ServiceLoader.load(LoggerFinder.class, cl == null ? ClassLoader.getSystemClassLoader() : cl);
+                    for (LoggerFinder lf : loader) {
+                        if (!ModuleLoggerFinder.class.equals(lf.getClass())) {
+                            finder = lf;
+                            break;
                         }
-                        if (result == null) {
-                            finder = result = defaultFinder.get();
+                    }
+                    if (finder == null) {
+                        finder = defaultFinder.get();
+                    }
+                } catch (Throwable ignore) {
+                    finder = defaultFinder.get();
+                }
+                // Process the queued loggers and drain the contents to new loggers from the finder found
+                final Iterator<Map.Entry<String, System.Logger>> iter = loggers.entrySet().iterator();
+                while (iter.hasNext()) {
+                    final Map.Entry<String, System.Logger> entry = iter.next();
+                    final System.Logger currentLogger = entry.getValue();
+                    // Check if we're currently a delegating logger which would need to be replaced with a real backing logger
+                    if (currentLogger instanceof DelegatingSystemLogger) {
+                        final DelegatingSystemLogger delegatingSystemLogger = (DelegatingSystemLogger) currentLogger;
+                        final System.Logger currentDelegate = delegatingSystemLogger.getDelegate();
+                        // If this is a queueing logger, we need to replace with a logger from the finder
+                        if (currentDelegate instanceof QueueingSystemLogger) {
```

Comment:
Thsi check is also not necessary because once we replace the delegate the map entry is being removed on line 126

Reply:
Same here too. Just feels safer even though it's likely to be the case.